src/scripts/upgrade.sh

```diff
@@ -0,0 +1,247 @@
+# Note: The audit log data would be restored immediately after the `iasql_*` modules are added, but not sure if we can do that with a "normal" SQL script or if we need to do something special there to get it in.
+# TODO: Also need to add an `UPGRADE` audit log type we can add in there to make it clear when prior logs are no longer fully "trustworthy" for the purposes of re-generating SQL statements from them.
+
+# Check if there's any tables in the metadata database, in case this is a first-time run
+HAS_IASQL_DATABASE=$(psql iasql_metadata -qtc "
+	SELECT table_name
+	FROM information_schema.tables
+	WHERE table_name = 'iasql_database';
+")
+
+if [ "$HAS_IASQL_DATABASE" != "" ]; then
+	# Get all user database names
+	USER_DATABASES=$(psql iasql_metadata -qtc "
+		SELECT pg_name
+		FROM iasql_database;
+	")
+	for DB in $USER_DATABASES
+	do
+		echo Checking $DB ...
+
+		# Get the engine and DB versions, figure out which is the newest and if that newest version is
+		# a beta release. *Always* re-run the upgrade on beta releases, and if not a beta release, run
+		# the upgrade only if the DB is an earlier version than the engine
+
+		ENGINE_VERSION=$(jq -r .version /engine/package.json | xargs) # xargs is the recommended way to 'trim' strings. Oh Bash...
+
+		DB_VERSION=$(psql $DB -qtc "
+			SELECT split_part(name, '@', 2) FROM iasql_module LIMIT 1;
+		" | xargs)
+
+		LATEST_VERSION=$(/engine/node_modules/.bin/semver $ENGINE_VERSION $DB_VERSION | tail -n 1 | xargs)
+
+		[[ "$LATEST_VERSION" == *"-beta" ]]
+		IS_BETA_VERSION=$?
+
+		[[ "$DB_VERSION" == "$LATEST_VERSION" ]]
+		DB_IS_LATEST=$?
+
+		SHOULD_UPGRADE=$IS_BETA_VERSION || ! $DB_IS_LATEST
+
+		if [[ $SHOULD_UPGRADE ]]; then
+			echo $DB to be upgraded!
+
+			# It's not possible to actually re-install the new modules without the engine, the RPC SQL is
+			# generated by Typescript code, so instead we're gonna grab the relevant info, stuff it into
+			# temporary files, then continue in Typescript hooking into the startup process, which will
+			# look for the `/tmp/upgrade/` path to decide if it needs to execute or not.
+
+			mkdir -p /tmp/upgrade/$DB
+
+			# Get list of modules that need to be re-installed
+			MODULE_LIST=$(psql $DB --csv -tc "
+				SELECT split_part(name, '@', 1) AS module FROM iasql_module;
+			")
+			echo $MODULE_LIST > /tmp/upgrade/$DB/module_list
+
+			for MODULE in $MODULE_LIST
+			do
+				if [[ "$MODULE" == "aws_account" ]]; then
+					# TODO: Figure out a more generic way of dealing with non-synced user data
+					psql $DB --csv -tc "
+						SELECT access_key_id, secret_access_key FROM aws_credentials;
+					" > /tmp/upgrade/$DB/creds
+          # Temporarily print the creds to see what's going on here
+          cat /tmp/upgrade/$DB/creds
```

Comment:
shall it be removed?

Reply:
Yep. The end of step 3 deletes `/tmp/upgrade/$DB` and after they're all done `/tmp/upgrade` is also deleted

Reply:
I think Yolanda means we should remove this `cat` call

Reply:
ohh, yes definitely should delete it :)